### PR TITLE
update ambiguous check

### PR DESCRIPTION
**Pull request by siddontang**
+ Use CheckReferAmbiguous to check fields ambiguous uniformly. 
+ Fix panic when adding hidden aggregate field.
+ Fix https://github.com/pingcap/tidb/issues/337

### CODE REVIEW

File: plan/plans/select_list.go [hunk 1]
```diff
@@ -140,6 +145,55 @@ func (s *SelectList) CloneHiddenField(name string, tableFields []*field.ResultFi
 	return false
 }
 
+// CheckReferAmbiguous checks whether an identifier reference is ambiguous or not in select list.
+// e,g, "select c1 as a, c2 as a from t group by a" is ambiguous,
+// but "select c1 as a, c1 as a from t group by a" is not.
+// For MySQL "select c1 as a, c2 + 1 as a from t group by a" is not ambiguous too,
+// so we will only check identifier too.
+// If no ambiguous, -1 means expr refers none in select list, else an index in select list returns.
+func (s *SelectList) CheckReferAmbiguous(expr expression.Expression) (int, error) {
```
Comment (shenli) [inline plan/plans/select_list.go]: I think the main purpose of this function is get the index of identifier from the select list.
GetIndex is a better name?

Comment (siddontang) [inline plan/plans/select_list.go]: No, the main purpose is to check ambiguous, getting index is just an incidental feature.

File: plan/plans/select_list.go [hunk 2]
```diff
@@ -140,6 +145,55 @@ func (s *SelectList) CloneHiddenField(name string, tableFields []*field.ResultFi
...
+	if _, ok := expr.(*expression.Ident); !ok {
+		return -1, nil
+	}
+
+	name := expr.String()
+	if strings.Contains(name, ".") {
```
Comment (qiuyesuifeng) [inline plan/plans/select_list.go]: use expression.IsQualified?

Comment (siddontang) [inline plan/plans/select_list.go]: Em....... I just know this function.

Comment (siddontang) [inline plan/plans/select_list.go]: Btw, I don't this function should be placed in expression package.

Comment (qiuyesuifeng) [inline plan/plans/select_list.go]: +1
maybe remove to field is better, and rename to  IsQualifiedField

File: rset/rsets/groupby.go [hunk 3]
```diff
@@ -93,30 +70,18 @@ func (r *GroupByRset) Plan(ctx context.Context) (plan.Plan, error) {
 			// use Position expression for the associated field.
 			r.By[i] = &expression.Position{N: position}
 		} else {
+			index, err := r.SelectList.CheckReferAmbiguous(e)
+			if err != nil {
+				return nil, errors.Errorf("Column '%s' in group statement is ambiguous", e)
+			} else if _, ok := aggFields[index]; ok {
```
Comment (qiuyesuifeng) [inline rset/rsets/groupby.go]: Can guarantee `lastIndex` is always valid?

Comment (siddontang) [inline rset/rsets/groupby.go]: if lastIndex is -1, aggFields[-1] is not ok.

File: rset/rsets/orderby.go [hunk 4]
```diff
@@ -140,9 +141,16 @@ func (r *OrderByRset) Plan(ctx context.Context) (plan.Plan, error) {
 				r.By[i].Expr = &expression.Position{N: position}
 			}
 		} else {
+			// Don't check ambiguous here, only check field exists or not.
```
Comment (qiuyesuifeng) [inline rset/rsets/orderby.go]: `No need to check ambiguous here` is better?

File: tidb_test.go [hunk 5]
```diff
@@ -1116,6 +1118,55 @@ func (s *testSessionSuite) TestDefaultFlenBug(c *C) {
...
+	mustExecSQL(c, se, "insert into t values (1,2), (2, 1)")
+
+	// Fix issue https://github.com/pingcap/tidb/issues/337
+	mustExecMatch(c, se, "select c1 as a, c1 as b from t order by c1", [][]interface{}{{1, 1}, {2, 2}})
+
+	mustExecMatch(c, se, "select c1 as a, t.c1 as a from t order by a desc", [][]interface{}{{2, 2}, {1, 1}})
+	mustExecMatch(c, se, "select c1 as c2 from t order by c2", [][]interface{}{{1}, {2}})
+	mustExecMatch(c, se, "select sum(c1) from t order by sum(c1)", [][]interface{}{{3}})
+
+	// TODO: now this test result is not same as MySQL, we will update it later.
```
Comment (qiuyesuifeng) [inline tidb_test.go]: How about use skip if will be updated later?

Comment (siddontang) [inline tidb_test.go]: No, we may forget to change this test if using skip.

File: rset/rsets/orderby.go (hunk 4 above)
Comment (siddontang) [inline rset/rsets/orderby.go]: Em, I think no need. Btw, I will refactor these code later too.

**Comment by qiuyesuifeng** [comment]
LGTM

File: plan/plans/select_list.go (hunk 2 above)
Comment (shenli) [inline plan/plans/select_list.go]: IsQualifiedIdent is better.

Comment (siddontang) [inline plan/plans/select_list.go]: Em, I prefer IsQualifiedName

File: plan/plans/select_list.go [hunk 6]
```diff
@@ -140,6 +145,55 @@ func (s *SelectList) CloneHiddenField(name string, tableFields []*field.ResultFi
...
+		// name is qualified, no need to check
+		return -1, nil
+	}
+
+	lastIndex := -1
+	// only check origin select list, no hidden field.
+	for i := 0; i < s.HiddenFieldOffset; i++ {
+		if s.ResultFields[i].Name != name {
```
Comment (shenli) [inline plan/plans/select_list.go]: Use case insensitive comparion?

File: plan/plans/select_list.go [hunk 7]
```diff
@@ -140,6 +145,55 @@ func (s *SelectList) CloneHiddenField(name string, tableFields []*field.ResultFi
...
+		}
+
+		if lastIndex == -1 {
+			// first match, continue
+			lastIndex = i
+			continue
+		}
+
+		// check origin name, e,g. "select c1 as c2, c2 from t group by c2" is ambiguous.
+		if s.ResultFields[i].ColumnInfo.Name.O != s.ResultFields[lastIndex].ColumnInfo.Name.O {
```
Comment (shenli) [inline plan/plans/select_list.go]: Why not use Name.L to do comparion?

**Comment by siddontang** [comment]
PTAL @shenli

**Comment by shenli** [comment]
LGTM

**Pull request merged**